BUG: Allow Series with same name with crosstab (#13279)

mroeschke opened the pull request:
- [x] closes #13279
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Before, passing two `Series` with the same name in `crosstab` raised a `KeyError` because of a block in `pivot_table` that tried to unstack a `MultiIndex` with duplicate names.

mroeschke committed:
BUG: Allow series with same in with crosstab (#13279)

mroeschke committed:
Add whatsnew

jreback commented on pandas/tools/pivot.py:
```diff
@@ -136,8 +136,14 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
     if table.index.nlevels > 1:
-        to_unstack = [agged.index.names[i] or i
-                      for i in range(len(index), len(keys))]
+        index_names = agged.index.names[:len(index)]
```
hmm is there a reason we don't just use ``i`` *always*?

mroeschke replied:
I tried this originally but some tests were failing because the MultiIndex names can also be ints. Looks like unstack first prioritizes index name if they aren't None and then defaults to index position by integer

```
In [19]: index = pd.MultiIndex.from_tuples([('one', 'a'), ('one', 'b'),
    ...: ...                                    ('two', 'a'), ('two', 'b'
    ...: )])

In [20]: s = pd.Series(np.arange(1.0, 5.0), index=index)

In [21]: s
Out[21]:
one  a    1.0
     b    2.0
two  a    3.0
     b    4.0
dtype: float64

In [22]: s.unstack(0)
Out[22]:
   one  two
a  1.0  3.0
b  2.0  4.0

In [24]: s.index.names= [1,0]

In [25]: s.unstack(0)
Out[25]:
0      a    b
1
one  1.0  2.0
two  3.0  4.0
```

jreback replied:
ahh, ok, so these should *always* use positionals then; you might need to change the downstream ``_get_level_values()`` (always positional) rather than ``get_level_values()`` (figures it out)

jreback replied:
actually we are just calling ``.unstack()`` so you can use just the positional then.

mroeschke replied:
Hmm downstream of this block, `get_level_values` is only used if `margins` is True, which in this case is False.

Also, the [unstack docs](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.unstack.html) mentions that level names can be passed, so it appears that `unstack` functionally tries to figure out whether a position or name is passed.

jreback replied:
I understand, but it is more clear simply to give all indexers (the level positioners) here rather than worry about determining if it has a name or not.

mroeschke replied:
Agreed, I'm just not sure how to force `unstack()` to recognize that I am passing along positional indexers and not names given that `unstack()` tries to figure it out (without breaking this functionality).

jreback replied:
yeah I would *just* always pass the actual indexers here. It will then just work.

if you can update we can get into 0.20.0.

mroeschke replied:
Sorry, its still not entirely clear to me how to pass the actual indexer (level positioners). Are level positioners always designated as ints?

Going back to the example where the index names are also ints:

```
In [58]: s
Out[58]: 
1    0
one  a    1.0
     b    2.0
two  a    3.0
     b    4.0
dtype: float64
```

If I want to unstack the `a,b,a,b` level in the 1st position, `unstack(level=1)` will unstack the level in the 0th position instead of the 1st.

jreback replied:
the indexer is the position of the level.

mroeschke replied:
But when calling `unstack(level=1)` in the example above, it is ambiguous whether to unstack  by level name or the level position since both are valid level identifiers in `unstack()` (but will to prioritize level name).

jreback replied:
right now we do a best guess interpretation. We can add a private argument to ``unstack`` which will handle only indexers (it would need to be passed thru to the underlying code), so a little bit of work.

IOW

``unstack([1, 0], _level_indexers=True)`` is purely unambiguous. This is something we could use from internal code (only).

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -1602,6 +1602,7 @@ Reshaping
 - Bug in ``DataFrame.pivot_table()`` where ``dropna=True`` would not drop all-NaN columns when the columns was a ``category`` dtype (:issue:`15193`)
 - Bug in ``pd.melt()`` where passing a tuple value for ``value_vars`` caused a ``TypeError`` (:issue:`15348`)
 - Bug in ``pd.pivot_table()`` where no error was raised when values argument was not in the columns (:issue:`14938`)
+- Bug in ``pd.crosstab()`` where passing two ``Series`` with the same name raised a ``KeyError`` (:issue:`13279`)
```
move to 0.20.2

jreback commented on pandas/tools/pivot.py:
```diff
@@ -136,8 +136,14 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
+        index_names = agged.index.names[:len(index)]
+        to_unstack = []
+        for i in range(len(index), len(keys)):
+            name = agged.index.names[i]
```
still like this to be more elegant.

mroeschke replied:
Sorry for being MIA on this PR. 

Unfortunately, things are getting busy on my end so I don't think I will have time implement a private argument to `stack()` (and probably `unstack()` as well) to indicate passing level or label indexers. I could add it on the original issue as the ideal solution.

I might be able to play around with this loop to make it cleaner.

jreback commented:
got lost, sorry. can you rebase / update.

jreback commented:
and move whatsnew to 0.21.0.

mroeschke committed:
Accidentally added some files

mroeschke commented:
Rebased and moved the whatsnew to 0.21.0. Apologies, I couldn't find a quick way to make the check more elegant yet. 

From our discussions before, you're correct that a cleaner solution would involve possibly including a private argument in `unstack()` to differentiate between positional and label indexers, but that solution would require more work.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -1723,4 +1723,4 @@ Other
 - Compat for 32-bit platforms for ``.qcut/cut``; bins will now be ``int64`` dtype (:issue:`14866`)
 - Bug in interactions with ``Qt`` when a ``QtApplication`` already exists (:issue:`14372`)
-- Avoid use of ``np.finfo()`` during ``import pandas`` removed to mitigate deadlock on Python GIL misuse (:issue:`14641`)
+- Avoid use of ``np.finfo()`` during ``import pandas`` removed to mitigate deadlock on Python GIL misuse (:issue:`14641`)
```
revert this file

jreback commented on pandas/core/reshape/pivot.py:
```diff
@@ -148,8 +148,14 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
     if table.index.nlevels > 1:
-        to_unstack = [agged.index.names[i] or i
-                      for i in range(len(index), len(keys))]
+        index_names = agged.index.names[:len(index)]
```
add a comment here about level names being in integer could be ambiguous (and ref the issue number about this; I think we have one)

jreback reviewed (approved):
minor comments. ping on green.

mroeschke committed:
Explain the logic of the fix

mroeschke committed:
Revert 0.20.0 whatsnew

mroeschke committed:
Revert "Revert 0.20.0 whatsnew"

This reverts commit 4b5302e0880f6d3914768941101660b9727839e2.

mroeschke committed:
Revert 0.20.0 whatsnew

mroeschke commented:
I couldn't find an exact, existing issue related to using integers with `unstack()` and a `MultiIndex`. #14969 is somewhat related but deals with `loc`. Should I make a new issue?

gfyoung commented:
> I couldn't find an exact, existing issue related to using integers with unstack() and a MultiIndex.

Go right ahead, though cross-reference #14969 just in case.

gfyoung commented on pandas/core/reshape/pivot.py:
```diff
@@ -148,8 +148,16 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
-        to_unstack = [agged.index.names[i] or i
-                      for i in range(len(index), len(keys))]
+        # If index_names are integers, determine whether the integers refer
+        # to the level position or name
```
Nit: Add a period at the end of this comment.

gfyoung commented:
@mroeschke : If you could reference your new issue in a comment for this patch and address my nit, I'll be happy to merge this afterwards.

mroeschke committed:
Add related issue number and add period

gfyoung commented:
Thanks @mroeschke !
